### ISSUE TITLE
Moved value is considered as held across an `await` in `async fn`

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust#87309]
I tried this code:

```rust
struct Ref<'a, T>(&'a T, *const () /* !Send */);

impl<'a, T> Ref<'a, T> {
    fn borrow(r: &'a T) -> Result<Self, ()> {
        Ok(Self(r, 0 as _))
    }
}

// Not `async fn` because `async fn` always captures lifetimes of arguments
fn accept_ref<'a, T>(
    _: Ref<'a, T>,
) -> impl core::future::Future<Output = ()> + Send + Sync + 'static {
    async {}
}

fn test<T: Send + Sync>(arg: T) {
    let fut = async move {
        match Ref::borrow(&arg) {
            Ok(r) => {
                // `r` is moved into `accept_ref` before `await`.
                let fut = accept_ref(r);
                Ok(fut.await)
            },
        //  - compiler thinks that `r` is dropped here, even though it's already moved 
        
            Err(()) => Err(arg),
        }
    };

    fn assert_send<T: Send>(_: T) {}

    assert_send(fut);
}
```
[[playground]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=51d54e549a0981de80e20493fd15238e)

I expected to see this happen: the code compiles.

Instead, this happened: the following compile error:

```text
error: future cannot be sent between threads safely
  --> src/lib.rs:32:5
   |
30 |     fn assert_send<T: Send>(_: T) {}
   |                       ---- required by this bound in `assert_send`
31 | 
32 |     assert_send(fut);
   |     ^^^^^^^^^^^ future created by async block is not `Send`
   |
   = help: within `impl Future`, the trait `Send` is not implemented for `*const ()`
note: future is not `Send` as this value is used across an await
  --> src/lib.rs:22:20
   |
19 |             Ok(r) => {
   |                - has type `Ref<'_, T>` which is not `Send`
...
22 |                 Ok(fut.await)
   |                    ^^^^^^^^^ await occurs here, with `r` maybe used later
23 |             },
   |             - `r` is later dropped here
```

### Meta

`rustc --version --verbose`:
```
rustc 1.55.0-nightly (240ff4c4a 2021-07-09)
binary: rustc
commit-hash: 240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4
commit-date: 2021-07-09
host: x86_64-unknown-linux-gnu
release: 1.55.0-nightly
LLVM version: 12.0.1
```

**Comment by tmandry** [comment]
Looks like an instance of #69663, potentially same as #57017.

@rustbot label: +AsyncAwait-Triaged

**Comment by jyn514** [comment]
Here is a reduced example:
```rust
use std::rc::Rc;

async fn non_trivial_drop() {
    let x = Rc::new(vec![0]);
    drop(x);
    async {}.await;
}


fn assert_send<T: Send>(_: T) {}

fn main() {
    assert_send(non_trivial_drop());
}
```

**Comment by eholk** [comment]
Does your example work with `-Zdrop-tracking`? This is exactly what drop tracking is meant to handle.

**Comment by jyn514** [comment]
@eholk yes, drop-tracking fixes it :) is there a tracking issue for that I can watch?

**Comment by jyn514** [comment]
hmm, https://github.com/rust-lang/rust/issues/69663 doesn't mention -Zdrop-tracking though - I would love to help get that onto stable :) let me know if there's anything I can do.

**Comment by eholk** [comment]
#69663 is the umbrella bug for all overly aggressive generator captures, but there are several distinct causes for them. The case you listed looks most like #57478, which is what I've been linked most of the drop tracking work to.

There are a couple of pending PRs fixing bugs in it, such as #96923 and #97029. Once #96923 merges, I think that fixes all the major bugs I know about so the next step will be to do a crater run to make sure we don't have any other bugs lurking around.

**Comment by eholk** [comment]
Actually, @jyn514, one issue I haven't been able to look at yet is `src/test/ui/async-await/issue-70935-complex-spans.rs`. With `-Zdrop-tracking` the error message is significantly worse. It still works, but I'd hate to regress the user experience like this. Is that something you'd be interested in looking at?

**Comment by jyn514** [comment]
Sure! let me take a look :)

**Comment by jyn514** [comment]
@eholk I'm not sure this is still an issue - when I add `// -Zdrop-tracking` at the top, the only things that change are the line numbers.

<details>

```diff
diff --git a/src/test/ui/async-await/issue-70935-complex-spans.rs b/src/test/ui/async-await/issue-70935-complex-spans.rs
index 2965a7e0654..be0c09bcb92 100644
--- a/src/test/ui/async-await/issue-70935-complex-spans.rs
+++ b/src/test/ui/async-await/issue-70935-complex-spans.rs
@@ -1,4 +1,5 @@
 // edition:2018
+// -Zdrop-tracking
 // #70935: Check if we do not emit snippet
 // with newlines which lead complex diagnostics.
 
diff --git a/src/test/ui/async-await/issue-70935-complex-spans.stderr b/src/test/ui/async-await/issue-70935-complex-spans.stderr
index db309938119..db6e17f3f5d 100644
--- a/src/test/ui/async-await/issue-70935-complex-spans.stderr
+++ b/src/test/ui/async-await/issue-70935-complex-spans.stderr
@@ -1,12 +1,12 @@
 error: future cannot be sent between threads safely
-  --> $DIR/issue-70935-complex-spans.rs:10:45
+  --> $DIR/issue-70935-complex-spans.rs:11:45
    |
 LL | fn foo(tx: std::sync::mpsc::Sender<i32>) -> impl Future + Send {
    |                                             ^^^^^^^^^^^^^^^^^^ future created by async block is not `Send`
    |
    = help: the trait `Sync` is not implemented for `Sender<i32>`
 note: future is not `Send` as this value is used across an await
-  --> $DIR/issue-70935-complex-spans.rs:15:11
+  --> $DIR/issue-70935-complex-spans.rs:16:11
    |
 LL |           baz(|| async{
    |  _____________-
@@ -14,9 +14,9 @@ LL | |             foo(tx.clone());
 LL | |         }).await;
    | |         - ^^^^^^ await occurs here, with the value maybe used later
    | |_________|
-   |           has type `[closure@$DIR/issue-70935-complex-spans.rs:13:13: 15:10]` which is not `Send`
+   |           has type `[closure@$DIR/issue-70935-complex-spans.rs:14:13: 16:10]` which is not `Send`
 note: the value is later dropped here
-  --> $DIR/issue-70935-complex-spans.rs:15:17
+  --> $DIR/issue-70935-complex-spans.rs:16:17
    |
 LL |         }).await;
    |                 ^
```

</details>

**Comment by jyn514** [comment]
oh uhh I'm dumb it's `// compile-flags:-Zdrop-tracking`

**Comment by eholk** [comment]
> @eholk yes, drop-tracking fixes it :) is there a tracking issue for that I can watch?

There is now: #77331

**Comment by StyMaar** [comment]
This issue seems to be solved now, the playground link compiles fine and everything listed in the tracking issue is marked as done (even though the tracking issue itself isn't)

**Comment by jmagnuson** [comment]
I had run into a similar issue, and bisected the fix to the 1.74.0 release which from the [changelog](https://releases.rs/docs/1.74.0/) and #97334, suggests that #107421 has ultimately resolved this.

**Comment by WaffleLapkin** [comment]
Given the above comments I'll go ahead and close this. @eholk I am assuming the drop tracking has enough tests and we don't need to do anything here?
